File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_네트워크.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43162
from collections import deque


def solution(n, computers):
    answer = 0
    visited = [0 for _ in range(n)]  # 각 노드 별 방문 여부
    d = deque()
    for i in range(n):
        if not visited[i]:  # 방문 안한 노드가 있으면
            answer += 1  # 네트워크 하나 추가
            visited[i] = 1  # 방문처리
            d.appendleft(i)  # 덱에 넣어둠
        while len(d) != 0:  # 덱이 빌 때까지 반복
            tmp = d.popleft()  # 덱을 큐로 사용. bfs
            for j in range(n):
                if computers[tmp][j] == 1 and not visited[j]:  # 연결 되어있고 연결된 노드에 방문 안했으면
                    visited[j] = 1  # 방문 처리
                    d.appendleft(j)

    return answer

File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_네트워크.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/43162
// C++에서는 배열 첫 선언 시 배열 크기를 지정해줘야 하는데, 이때 크기로 변수를
// 사용하면 안된다 이 부분에서 헤맸었음

#include <queue>
#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<vector<int>> computers) {
  int answer = 0;
  bool visited[200] = {0};  // 각 노드 당 방문 여부 확인 위함
  queue<int> q;
  for (int i = 0; i < n; i++) {
    if (!visited[i]) {                 // 노드 방문 안했을 경우
      answer += 1;                     // 네트워크 수 추가
      visited[i] = 1;                  // 방문처리
      q.push(i);                       // 큐에 노드 추가
      while (!q.empty()) {             // 큐가 빌 때까지
        int current_node = q.front();  // 맨 앞 요소 가져옴
        q.pop();
        for (int j = 0; j < n; j++) {
          if (computers[current_node][j] == 1 &&
              !visited[j]) {  // 연결되어있고 방문 아직 안했을 경우
            visited[j] = 1;  // 방문 여부 갱신
            q.push(j);       // 큐에 노드 추가
          }
        }
      }
    }
  }
  return answer;
}